rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,468 @@
+in the backup snapshot taken before an upgrade, rolling back to the previous
+snapshot will automatically remove the guard.
+
+**7.x Minimize data loss when creating new indices**
+
+Preventing data loss from mixed 7.x versions, introduces two additional steps
+to rollback to a previous version without a snapshot:
+1. (existing) Point the `.kibana` alias to the previous Kibana index `.kibana_n-1`
+2. (existing) Delete `.kibana_n`
+3. (new) Enable writes on `.kibana_n-1`
+4. (new) Delete the dummy "version lock" document from `.kibana_n-1`
+
+Since our documentation and server logs have implicitly encouraged users to
+rollback without using snapshots, many users might have to rely on these
+additional migration steps to perform a rollback. Since even the existing
+steps are error prone, introducing more steps will likely introduce more
+problems than what it solves.
+
+<details>
+  <summary><s>Minimize data loss with mixed 7.x versions</s></summary>
+  
+  1. All future versions of Kibana 7.x will use the `.kibana_saved_objects`
+   alias to locate the current index. If `.kibana_saved_objects` doesn't
+   exist, newer versions will fallback to reading `.kibana`.
+  2. All future versions of Kibana will locate the index that
+     `.kibana_saved_objects` points to and then read and write directly from
+     the _index_ instead of the alias.
+  3. Before starting a migration:
+    1. Write a new dummy "version lock" document to the `.kibana` index with a
+      `migrationVersion` set to the current version of Kibana. If an outdated
+      node is started up after a migration was started it will detect that
+      newer documents are present in the index and refuse to start up.
+    2. Set the outdated index to read-only. Since `.kibana` is never advanced,
+      it will be pointing to a read-only index which prevent writes from
+      6.8+ releases which are already online.
+</details>
+
+## 4.5 Changes in 8.0
+1. Re-use the same index for minor upgrades.
+2. Only allow breaking field mapping changes (which requires creating a new
+   index) in a major.
+3. All update operations should use optimistic concurrency control.
+
+### Migration algorithm (8.0):
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed.
+2. If there are documents with newer `migrationVersion` numbers, exit Kibana
+   with a fatal error.
+3. If there are outdated documents, migrate these in batches:
+   1. Read a batch of outdated documents from `.kibana_n`.
+   2. Transform documents by applying the migration functions.
+   3. Update the document batch in the same index using optimistic concurrency
+      control. 
+   4. If a batch fails due to an update version mismatch, repeat the entire
+      process.
+4. Once all documents are up to date, the migration is complete and Kibana can
+   start serving traffic.
+
+Advantages:
```

Comment:
Wanted to add something here, since we just hit it with a customer.

- Avoids an issue where Kibana can not start due to being unable to create a new index. An example of this is when the cluster is over [cluster.max_shards_per_node](https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-cluster.html#cluster-shard-limit)

Reply:
Added in 39ec98c